### PR TITLE
fix(schema): stop an illegal Arrow Map entries declaration from being stored or compared (fixes #13549)

### LINKED ISSUE #13549: An accelerator that already persisted a nullable MAP entries declaration keeps it, so the ingress correction never reaches it
## What

#13495 corrects a `MAP` column's `entries` nullability at every Arrow decode point, so a producer that
declares `entries` nullable — which the Arrow map layout forbids — no longer hands a column into the
plan that fails in whichever kernel first rebuilds it. That correction is at the *ingress*. An
acceleration that has already persisted the non-conforming declaration is not reached by it, and two
mechanisms keep it that way.

## Where

- `crates/cayenne/src/provider/sink.rs:146` — the sink casts every incoming batch to the stored table
  schema (`arrow_tools::record_batch::try_cast_to`). A batch corrected at ingress is therefore cast
  *back* to a stored `Map(entries nullable)` declaration on the way in.
- `crates/cayenne/src/cayenne_catalog.rs:1351` — a difference between the source schema and the stored
  one that is a nullability tighten leaves the stored schema canonical, so a source that now reports
  the conforming declaration does not update the table.

## Why it matters

The stored declaration is the one every read of that table is planned against, so the column stays
one no kernel can rebuild, reporting `MapArray entries cannot contain nulls` whether or not a null is
involved. Nothing surfaces the state: the table loads, reports healthy, and fails only when a query
touches the map column.

The exposure is narrow — it needs an accelerated table created from a non-conforming producer before
the ingress fix — but it does not self-heal, because the two mechanisms above are what keep a stored
schema stable across restarts.

## Shape of a fix

Whatever corrects this has to distinguish "the source relaxed a nullability constraint" (which the
retention rule exists to ignore) from "the stored declaration violates the Arrow map layout and can
never be read". The second is not a schema *evolution* at all: relabelling `entries` non-nullable
touches no buffer, so it can be applied to the stored schema without rewriting a single file.

Found while fixing #13495.

### PR DESCRIPTION
**Pull request by claudespice**
## Summary

The Arrow map layout forbids a `MAP`'s `entries` field from being nullable, and `MapArray::try_new` refuses one — but nothing enforces it at decode, so a producer that declares it that way hands over a column that decodes cleanly and then fails in whichever kernel first rebuilds it, reporting `MapArray entries cannot contain nulls` whether or not a null is involved.

spiceai/spiceai#13495 corrected that at every Arrow decode point. This is the half that correction cannot reach: a Cayenne accelerator that has **already persisted** the non-conforming declaration. The stored declaration is what every read of the table is planned against and what the write sink casts each incoming batch to, and the table does not recover on its own, because the rule that keeps a stored schema canonical across a nullability difference is exactly what holds the illegal declaration in place.

**One correction to the issue's premise, measured rather than assumed.** #13549 says the column "stays one no kernel can rebuild". I tested that directly — a table holding rows written under the forbidden declaration, reopened against the repaired one — and it **scans fine**, on both the inline corpus and after a checkpoint flushes it to a Vortex file, under projection, concat, sort and join alike. So the unreadable-column shape is not what a Cayenne scan does with it, and this PR should not be read as fixing that.

What the illegal declaration demonstrably breaks is every schema **comparison**, and that is the failure this PR is worth landing for. It is not hypothetical: removing the normalization reproduces `incompatible schema change detected on the CDC stream … which is not a lossless widening. on_schema_change: fail is set` — a dataset that stops replicating over a schema that never changed.

## Changes

**`crates/cayenne/src/cayenne_catalog.rs`** — two ends of the metastore boundary:

- `create_table` conforms `options.schema` on entry, so the declaration the table is persisted under, compared against and evolved from is the same conforming one. It sits at the entry rather than at the serializer because `options.schema` is compared by `configuration_matches` — which is not the widening classifier and does not normalize — long before anything is persisted.
- `get_table` conforms the decoded stored schema, which heals a table persisted before this on its next load. Both are metadata-only — nullability lives in the type rather than in any buffer — so no file is rewritten.

**`crates/arrow_tools/src/type_rewrite.rs`** — the same declaration also has to stop reading as a *difference*. `normalize_for_comparison` applies the map rule alongside the dictionary unwrapping that was already normalized away for the same reason, and both schema comparisons now go through it:

- `schema_evolution::classify` — no widening rule can reconcile the two declarations, so once the stored one conforms, a source that keeps sending the forbidden one differs from it on **every** comparison and reads as an unevolvable change: per CDC batch, at restart through `FederatedTable::new_with_schema_change_policy`, and at the accelerator's restart-time reclassification, where it also reached `update_table_schema` and persisted the declaration verbatim. Under `on_schema_change: fail` that stops a dataset replicating over a schema that never changed. Normalizing here also means no `WideningPlan::evolved_schema` can carry the declaration into whatever persists or installs it, which covers the live Cayenne swap and every catalog schema write without a guard at each one.
- `narrowing_schema_cast_changes` (`crates/runtime-table/src/accelerated/sink/table.rs`) runs the same "normalize both sides, then compare" once per insert plan and had only the dictionary half, so the pair landed in `narrowed` and warned about a silent narrowing on every insert.

The rewrite is one bit on one type node, and the two declarations describe an identical physical layout — one of them a layout Arrow refuses to materialize at all — so no distinction any engine could hold is collapsed. Every real difference still reaches the classifier; a renamed entries child is still refused.

**`docs/cayenne/cayenne.md`** — the `Map` round-trip paragraph now says which declaration a map is stored under, and why comparisons are normalized.

## Performance impact

None measurable. `classify` already made exactly two `apply_rules` calls; this adds one rule to the existing slice, which is one indirect call and one discriminant compare per type node, no new allocation, and unchanged refcount sharing of fields the rules do not touch. `conforming_schema` returns the same `Arc` when nothing needs changing, and both Cayenne call sites are DDL/open paths, not per-batch. One caveat is filed as spiceai/spiceai#13693: the CDC fast path (`cdc_data_schema_matches`) asks the *engine* rewrite rules, which this layout rule deliberately does not join, so for the schemas this fix targets it misses per sub-batch and the full comparison runs. That work is not new — it previously ran and terminated in a rejection — but it is now an ongoing steady state rather than a one-off before a stop.

## Test plan

- [x] Added regression tests for the persist and read halves (`create_table_persists_a_conforming_map_entries_declaration` asserts the **stored bytes**, so `get_table`'s repair cannot mask a broken create path; `get_table_repairs_a_stored_nonconforming_map_entries_declaration` writes legacy bytes directly)
- [x] Added regression tests for the comparison halves: `an_entries_nullability_difference_is_not_a_change` and `an_evolved_schema_never_carries_the_forbidden_declaration` (arrow_tools), `cdc_schema_evolution_accepts_a_nonconforming_map_entries_declaration` and `a_map_entries_declaration_the_layout_forbids_is_not_a_narrowing` (runtime-table), `schema_evolution_live_installs_a_conforming_map_entries_declaration` (cayenne, asserting the metastore write and the in-memory swap separately)
- [x] Added edge case tests: a nested map reached through a `Struct`; a schema with no `MAP` round-tripping unchanged including field and schema metadata; a renamed entries child still refused as a real difference; `configuration_matching_does_not_normalize_a_map_entries_declaration`, which pins the property the `create_table` placement depends on
- [x] Added `legacy_inline_map_data_is_readable_once_the_declaration_is_repaired`: rows written through the ordinary insert path, the stored blob then swapped for the same rows re-serialized under the pre-conformance declaration, read back from the inline corpus and again after a checkpoint flushes to Vortex. It is a forward guard, **not** a guard on the repair — removing the repair makes the stored declaration agree with the blob again, so there is no disagreement left to read; its doc comment says so
- [x] Every guard neuter-verified: removing each site fails the tests naming it, and removing the shared normalizer fails tests in all three crates. The two data-bearing tests above are the stated exception
- [x] `make lint` passes (workspace-wide `cargo fmt --all -- --check` + full clippy — not a per-crate scoped invocation)
- [x] `make test` passes — 12051 tests, 0 failures (one flaky unrelated to this diff, `runtime-rate-control leased::tests::pre_lease_lands_in_next_window_and_promotes_on_roll`, passed on retry)
- [x] `make signoff` run locally after each push, on the current HEAD (`Attestation` check is green)

## Review gate

- Adversarial review: **skipped** — `codex` exited 1: "Your workspace is out of credits. Ask your workspace owner to refill in order to continue."; `grok` not installed. Two earlier `codex` passes on this branch did run and shaped what shipped: job `review-mtdutx5u-g1834t` (1 high finding — the invariant was confined to create/read and schema evolution could reinstate the declaration; fixed, and tracing it surfaced that a non-conforming source would then classify `Incompatible` on every batch, which the same fix covers) and job `review-mtdw6i31-l25q46` (1 high, 1 medium — restart-time classification and the accelerator's restart-time evolution were both uncovered; both fixed by moving the normalization into `classify`, which is that review's own recommended resolution). Re-attempted on each later HEAD and refused each time, so the last two commits — the cleanup pass and the data-bearing test — carry no adversarial review. Refilling the workspace's Codex credits would restore that gate; a reviewer should weigh those two commits accordingly, though both are test/doc changes with no production delta beyond the shared normalizer already reviewed.
- `/security-review`: no findings. Run three times as the diff changed; the last pass specifically bounded the `narrowing_schema_cast_changes` change, since relaxing a safety check's comparison is the one shape here that could hide data loss, and confirmed every pair it newly equates is physically identical with one side a declaration Arrow refuses to materialize.
- `/simplify`: applied — the shared normalizer, the extracted `deserialize_schema_ipc_base64` (the inverse of the serializer it now sits beside), a deleted hand-rolled test oracle that was weaker than the assertions beside it, and reuse of existing test helpers. Light checks re-run green.
- Copilot raised that the tests stopped at the `SchemaRef` and never exercised a legacy table holding map **data**; @grokspice confirmed the gap and filed spiceai/spiceai#13700. I could run the cayenne suite, so I answered it here instead of deferring: the data-bearing case is now covered by a test, and the answer is that it reads correctly on both tiers — which refutes both candidate outcomes #13700 was written around, and is the measured correction to the issue's premise noted in the Summary.
- Deferred, filed rather than widened: spiceai/spiceai#13693 (CDC fast-path miss) and spiceai/spiceai#13694 (the acceleration snapshot store, the one persistence path this fix's reasoning does not reach).

Fixes #13549

## Checklist

- [x] Data correctness: the change is metadata-only, rewrites no file, and collapses no distinction any engine could hold
- [x] No `.unwrap()`; tests use `.expect()` with a message
- [x] `docs/cayenne/cayenne.md` reviewed and updated (behavior change to how a `Map` is stored)

### CODE REVIEW

**Comment by claudespice** [comment]
@copilot review

File: crates/cayenne/src/cayenne_catalog.rs [hunk 1]
```diff
@@ -2329,28 +2342,14 @@ impl MetadataCatalog for CayenneCatalog {
...
-
-                    let schema = Arc::new(schema);
+                    // A schema persisted before `create_table` began conforming its input —
+                    // or by a Spice that predates it — still declares its `MAP` entries the way
+                    // its producer did, and nothing else will ever repair it: the retention rule
+                    // that keeps a stored schema canonical across a nullability difference is
+                    // what holds it in place. Repairing it on the way out heals such a table on
+                    // its next load without rewriting a single file.
+                    let schema =
+                        conforming_schema(Arc::new(deserialize_schema_ipc_base64(&schema_json)?));
```
[the remark this answers is not in the document]
Comment (grokspice) [reply crates/cayenne/src/cayenne_catalog.rs:2352]: You are right that the tests stop at the `SchemaRef`, and the mechanism you name for why that is not enough holds up. I traced it rather than taking it on faith, and it is filed as **#13700** rather than folded in here.

**The decode does read the file's own declaration, not the table's.** `crates/cayenne/src/provider/table.rs:658`:

```rust
let reader = arrow::ipc::reader::StreamReader::try_new(std::io::Cursor::new(ipc_bytes), None)?;
```

The `None` is the projection; the schema comes from the stream header. So for an inline blob written before the conformance, the decoded batch declares `entries` nullable while the repaired `table_schema` declares it non-nullable — the two disagree on precisely the bit this PR corrects. That is a real, untested path, and "the repair costs no file rewrite" is exactly what makes it reachable: the bytes keep their old declaration by design.

**Both ways it can go are untested, and they are not the same bug.** Either the IPC decode refuses outright — leaving the column unreadable for a new reason — or it succeeds, since arrow-rs builds arrays through `ArrayData`/`make_array` rather than `MapArray::try_new`, and the mismatch surfaces later in whichever kernel first rebuilds the map against the target schema, as the original `MapArray entries cannot contain nulls`. Which one happens decides whether this PR is sufficient for a data-bearing legacy table or only for an empty one. Vortex-backed rows want the same question asked of their own adaptation.

**Why it is #13700 and not a commit here.** Answering it means asserting a *scan*, not a schema: create the table, write map rows, restore the pre-conformance `schema_json` (the `store_legacy_declaration` helper on this PR already does that half), reopen, and query the column — across the inline tier and file-backed storage. If the mismatch does surface, the fix is a read-path adaptation for this bit, the way widening already adapts old files, and that is a change to the scan path rather than to the catalog layer this PR touches. Writing that test without being able to run the cayenne suite locally would be shipping an assertion I have not seen fail for the right reason, which is worse than filing the gap where it can be worked deliberately.

What this PR does still stands on its own and is worth landing as-is: it stops the illegal declaration being persisted, and it heals the declaration of an existing table on load — which is a precondition for any read-path repair, since nothing downstream can adapt *to* a target that is itself illegal. #13700 carries the rest, with the reproduction shape above and the two candidate outcomes written down.

**Review by lukekim — APPROVED** [review]
